src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -27,178 +25,78 @@ public class CompilerCache : ICompilerCache
         /// Initializes a new instance of <see cref="CompilerCache"/> populated with precompiled views
         /// discovered using <paramref name="provider"/>.
         /// </summary>
-        /// <param name="assemblyProvider">The <see cref="IAssemblyProvider"/> that provides assemblies
-        /// for precompiled view discovery.</param>
+        /// <param name="razorFileInfoCollections">The sequence of <see cref="RazorFileInfoCollection"/> that provides
+        /// information for precompiled view discovery.</param>
         /// <param name="loaderContextAccessor">The <see cref="IAssemblyLoadContextAccessor"/>.</param>
         /// <param name="optionsAccessor">An accessor to the <see cref="RazorViewEngineOptions"/>.</param>
-        public CompilerCache(IAssemblyProvider assemblyProvider,
-                             IAssemblyLoadContextAccessor loadContextAccessor,
-                             IOptions<RazorViewEngineOptions> optionsAccessor)
-            : this(GetFileInfos(assemblyProvider.CandidateAssemblies),
-                  loadContextAccessor.GetLoadContext(RazorFileInfoCollectionType.Assembly),
+        public CompilerCache(
+            IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
+            IAssemblyLoadContextAccessor loadContextAccessor,
+            IOptions<RazorViewEngineOptions> optionsAccessor)
+            : this(razorFileInfoCollections,
+                  loadContextAccessor.GetLoadContext(typeof(CompilerCache).GetTypeInfo().Assembly),
                   optionsAccessor.Options.FileProvider)
         {
         }
 
-        internal CompilerCache(IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
-                               IAssemblyLoadContext loadContext,
-                               IFileProvider fileProvider)
+        internal CompilerCache(
+            IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
+            IAssemblyLoadContext loadContext,
+            IFileProvider fileProvider)
```

Comment:
We had a similar discussion in the past and at that point we decided it was saner to initialize the constructor to avoid having to perform the `EnsureInitialized` check as part of `GetOrAdd`. 

We have a test that verifies in the scenario where a View is never returned, the cache is never initialized. https://github.com/aspnet/Mvc/blob/dev/test/WebSites/RazorCompilerCacheWebSite/Services/CustomCompilerCache.cs#L20

Granted it does assume that nobody except RazorViewEngine attempts to resolve this service, but my guess is that's the most common scenario.